crates/services/executor/src/executor.rs

```diff
@@ -486,137 +375,220 @@ where
+    {
+        let amount_to_mint = if coinbase_contract_id != ContractId::zeroed() {
+            data.coinbase
+        } else {
+            0
+        };
+        let block_height = *block.header.height();
+
+        let coinbase_tx = Transaction::mint(
+            TxPointer::new(block_height, data.tx_count),
+            input::contract::Contract {
+                utxo_id: UtxoId::new(Bytes32::zeroed(), 0),
+                balance_root: Bytes32::zeroed(),
+                state_root: Bytes32::zeroed(),
+                tx_pointer: TxPointer::new(BlockHeight::new(0), 0),
+                contract_id: coinbase_contract_id,
+            },
+            output::contract::Contract {
+                input_index: 0,
+                balance_root: Bytes32::zeroed(),
+                state_root: Bytes32::zeroed(),
+            },
+            amount_to_mint,
+            *self.consensus_params.base_asset_id(),
+            gas_price,
+        );
+
+        self.execute_transaction_and_commit(
+            block,
+            storage_tx,
+            data,
+            MaybeCheckedTransaction::Transaction(coinbase_tx.into()),
+            gas_price,
+            coinbase_contract_id,
+        )?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument(skip_all)]
+    /// Execute dry-run of block with specified components
+    fn dry_run_block<TxSource>(
+        mut self,
+        components: Components<TxSource>,
+    ) -> ExecutorResult<(PartialFuelBlock, ExecutionData)>
+    where
+        TxSource: TransactionsSource,
+    {
+        let empty_block = PartialFuelBlock::new(components.header_to_produce, vec![]);
+        let mut data = ExecutionData::new();
+
+        let block_height = *empty_block.header.height();
+        let da_block_height = empty_block.header.da_height;
+        let l1_transactions =
+            self.get_relayed_txs(block_height, da_block_height, &mut data)?;
+
+        let mut block_storage_tx = self
+            .block_storage_tx
+            .read_transaction()
+            .with_policy(ConflictPolicy::Overwrite);
+
+        let skip_failed_txs = false;
+        let partial_block = self.process_l1_and_l2_txs(
+            empty_block,
+            &components,
+            l1_transactions,
+            &mut block_storage_tx,
+            &mut data,
+            skip_failed_txs,
+        )?;
+
+        let changes_from_execution = block_storage_tx.into_changes();
+        self.block_storage_tx
+            .commit_changes(changes_from_execution)?;
+
+        data.changes = self.block_storage_tx.into_changes();
+        Ok((partial_block, data))
+    }
+
+    fn process_l1_and_l2_txs<T, TxSource>(
+        &self,
+        mut block: PartialFuelBlock,
+        components: &Components<TxSource>,
+        l1_transactions: Vec<CheckedTransaction>,
+        storage_tx: &mut T,
+        data: &mut ExecutionData,
+        skip_failed_txs: bool,
+    ) -> ExecutorResult<PartialFuelBlock>
```

Comment:
Would it be feasible to break this up into something like `process_l1` and `process_l2`, and simply call these two functions sequentially?

Reply:
Yeah. It's worth looking at a bit more. It does seem like a smell to me as well. I included them together because `produce` and `dry_run` used the exact same code block, so it was a nice way to group those together. I think there probably exists a cleaner solution.